### PR TITLE
ggplot histograms vertical breaks fix

### LINKED ISSUE #702: problem with ggplot histograms
there's some error in the ggplot histograms when using the fill option:

Screenshot 2023-06-30 at 1 38 40 p m

from what I understand, fill should display one histogram on top of another with different colors. since these are histograms, they should not have vertical color breaks (beginning from the bottom) where they start with color A, then color B and then they are color A again. since the value in the Y axis is a count. see: https://stackoverflow.com/questions/22216350/overlay-histograms-in-r

hist

### PR DESCRIPTION
**Pull request by bbeat2782**
## Describe your changes

Fixed vertical color breaks in histograms by changing the method of getting bin width.

## Issue number

Closes #702 

## Checklist before requesting a review

- [x] Performed a self-review of my code
- [x] Formatted my code with [`pkgmt format`](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#linting-formatting)
- [ ] Added [tests](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#testing) (when necessary).
- [x] Added [docstring](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#documenting-changes-and-new-features) documentation and update the [changelog](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#changelog) (when needed)

----
:books: Documentation preview :books:: https://jupysql--711.org.readthedocs.build/en/711/

### CODE REVIEW

**Comment by bbeat2782** [comment]
With the current change in this PR, I was able to make the following fix. 

Screen Shot 2023-07-05 at 8 46 43 AM

### Summary
Used `bin_size` returned by `_histogram()` instead of calculating bin width with `bins[-1] - bins[-2]`.

### Cause of the error
After the query below,
Screen Shot 2023-07-05 at 12 55 25 PM

the last two bins are far away from each other more than the other bins. (For other bins, the distance is 180. For the last two, it is 360.)



Thus, using `bins[-1] - bins[-2]` to get width of bins during plotting causes the error described in #702.
Screen Shot 2023-07-05 at 12 58 03 PM

As you can see below, a similar error occurs in the test plot images.



### Solution
`_histogram()` in `plot.py` computes `bin_size` with `range_ / bins` if `min_` and `max_` are numeric values. I thought using the `bin_size` computed here is more appropriate than `bins[-1] - bins[-2]`, and this is the change I made in the `plot.py`.

**Comment by edublancas** [comment]
@bbeat2782 did you validate (at least some) of the plots with R's ggplot?

**Comment by bbeat2782** [comment]
> @bbeat2782 did you validate (at least some) of the plots with R's ggplot?

@edublancas Yes, I was able to replicate all the histograms' shapes through R ggplot2 using the bin create method we use in the `plot.py` file. I didn't try all the color combinations because the colors are consistent between old and new images.

File: src/sql/plot.py [hunk 1]
```diff
@@ -306,14 +306,14 @@ def _are_numeric_values(*values):
     return all([isinstance(value, (int, float)) for value in values])
 
 
-def _get_bar_width(ax, bins):
+def _get_bar_width(ax, bins, bin_size):
     """
```
Comment (neelasha23) [inline src/sql/plot.py:310]: Please add a docstring here with brief explanation of each parameter

**Comment by neelasha23** [comment]
1. For the [histogram_numeric_categorical_combined_custom_multi_color](https://github.com/ploomber/jupysql/pull/711/files#diff-2ebd1d0a31d4ff5986d7e7e1bb7dcf399f6d924c1bc6dc548a6d6cb7ebdbe51d), the bar width of `color` looks similar in the new image, but the bar of `carat` is much narrower. Is this expected? Same for many other images.
2. In this [src/tests/integration/baseline_images/test_questDB/custom_engine_histogram.png](https://github.com/ploomber/jupysql/pull/711/files#diff-51660bcfefc2452bc4c5e14bbbb664710b6b1bad4e0a68ba9d4c20c92c98bceb), the blue overlay bars seem gone in the new image. 
3. Can we add an example in the [docs](https://jupysql.ploomber.io/en/latest/user-guide/ggplot.html)

**Comment by bbeat2782** [comment]
@neelasha23 
> 1. For the [histogram_numeric_categorical_combined_custom_multi_color](https://github.com/ploomber/jupysql/pull/711/files#diff-2ebd1d0a31d4ff5986d7e7e1bb7dcf399f6d924c1bc6dc548a6d6cb7ebdbe51d), the bar width of `color` looks similar in the new image, but the bar of `carat` is much narrower. Is this expected? Same for many other images.
> 2. In this [src/tests/integration/baseline_images/test_questDB/custom_engine_histogram.png](https://github.com/ploomber/jupysql/pull/711/files#diff-51660bcfefc2452bc4c5e14bbbb664710b6b1bad4e0a68ba9d4c20c92c98bceb), the blue overlay bars seem gone in the new image.

As far as I know, there shouldn't be overlapping areas when plotting a histogram with one variable. Thus, for both cases (`carat` and `bill_length_mm`), I believe those overlapping areas should be gone.

And from my understanding, the narrower bar widths are correct because the larger bar widths are due to considering `bins[-1] - bins[-2]` as bar width when `bins[-1] - bins[-2]` is larger than distances between other two consecutive bins. 

If any of the things I comment on is wrong, please correct me. 


> 3. Can we add an example in the [docs](https://jupysql.ploomber.io/en/latest/user-guide/ggplot.html)

Could you please elaborate more on what kind of example you are asking? I'm a little bit confused on this point because it seems like the link already has examples of plotting different histograms. Thanks.

**Comment by neelasha23** [comment]
> Could you please elaborate more on what kind of example you are asking? I'm a little bit confused on this point because it seems like the link already has examples of plotting different histograms. Thanks.

I don't see one similar to the examples in the testcases. (with overlay). Maybe it's because of the dataset

@bbeat2782

**Review by yafimvo — CHANGES_REQUESTED** [review]
I think the change makes sense and it looks good. However, there are three cases where we calculate the width, and we might calculate it differently for some of them.

For example:

```
%sqlplot histogram --table penguins.csv --column body_mass_g --bins 300
```

here for example we have 1 color without fill. In this case, when the bars are overlapping each other, it is actually easier to understand the chart.

image

**Comment by bbeat2782** [comment]
@yafimvo 
> ```
> %sqlplot histogram --table penguins.csv --column body_mass_g --bins 300
> ```
> 
> here for example we have 1 color without fill. In this case, when the bars are overlapping each other, it is actually easier to understand the chart.
> 
> image

To check what I understand is correct, are you saying that when a user plots a histogram with 1 color and without fill (no stacked), the bars of histograms should overlap with each other no matter the number of bins?

> However, there are three cases where we calculate the width, and we might calculate it differently for some of them.

And are these the other two cases you mentioned?

Screen Shot 2023-07-06 at 9 01 17 AM
Screen Shot 2023-07-06 at 9 01 31 AM

**Comment by bbeat2782** [comment]
> 3\. fitted (example: bar 1 begins at x=0 and ends at x=1, while bar 2 begins at x=1)  - we want this! strictly speaking, there is a slightly small overlap but that's fine, because we want the histogram to look "continuous"

@edublancas 

It clarifies that we want to have **3. fitted** version of histograms. 

The current implementation gives **1. overlapping** histograms which produces the vertical color breaks.
My implementation gives the following, which is the **2. spaced** version. 
Screen Shot 2023-07-07 at 11 50 55 AM

So to confirm my understanding, are you saying that we would like to have all bins stand right next to each other even if it means the bins in the same histogram have different widths like the following image?
Screen Shot 2023-07-07 at 11 45 29 AM

**Comment by edublancas** [comment]
> So to confirm my understanding, are you saying that we would like to have all bins stand right next to each other (so that they look "continuous") even if it means the bins in the same histogram have different widths like the following image?

If using `%sqlplot histogram`, I don't think there's a way for bins to have different width. but I'm unsure if this is a possible scenario with the `ggplot` API. is this possible? If so, it'd say let's stick with what ggplot in R does, as we want to mimic their API

**Comment by bbeat2782** [comment]
> is this possible? If so, it'd say let's stick with what ggplot in R does, as we want to mimic their API

Just checked that having different bins widths is possible by passing the `breaks` argument. But this isn't implemented in our `geom_histogram.py`. So I will start from here.

Screen Shot 2023-07-07 at 12 14 12 PM

**Comment by bbeat2782** [comment]
>  is this possible? If so, it'd say let's stick with what ggplot in R does, as we want to mimic their API

By default, a histogram plotted by ggplot in R contains some whitespaces between bins if the number of bins is large and some bins do not contain a data point.
Screen Shot 2023-07-07 at 1 20 01 PM


We can change the width of bins and remove the whitespaces by passing in the `breaks` argument, which we don't have in our `geom_histogram.py`
Screen Shot 2023-07-07 at 1 14 29 PM

@edublancas So I will work from there. Thanks for the clarification.

**Comment by edublancas** [comment]
ok, to ensure I understand what you'll do:


you'll mimic what ggplot in R does (this means that in some cases there will be whitespace if data is missing in certain data ranges).

you will not implement `breaks`

**Comment by bbeat2782** [comment]
@edublancas 
> you'll mimic what ggplot in R does (this means that in some cases there will be whitespace if data is missing in certain data ranges).

Yes

> you will not implement `breaks`

I was going to implement `breaks` since it's in ggplot in R, but if we don't want it, I won't.

**Comment by edublancas** [comment]
let's tackle `breaks` in another PR. I'll open an issue

**Comment by bbeat2782** [comment]
@yafimvo 

After clarifying the following with edublancas 
> mimic what ggplot in R does (this means that in some cases there will be whitespace if data is missing in certain data ranges).

do you think we still need to tackle the 3 plots you mentioned from your first comment?

**Comment by bbeat2782** [comment]
@neelasha23 
Could you take a look at the changes I made based on your comment? Thanks!

File: doc/user-guide/ggplot.md [hunk 2]
```diff
@@ -185,6 +185,12 @@ We can show a histogram of multiple columns by setting `x=['cut', 'color']`
 (ggplot("diamonds", aes(x=["cut", "color"])) + geom_histogram())
 ```
 
+Showing multiple columns also work with a mix of categorical and numerical columns
```
Comment (neelasha23) [inline doc/user-guide/ggplot.md:188]: We can also plot histograms for a combination of categorical and numerical columns.

Comment (bbeat2782) [reply doc/user-guide/ggplot.md:188]: Just to confirm, is this suggesting changing the wordings?

Comment (neelasha23) [reply doc/user-guide/ggplot.md:188]: yes

Comment (bbeat2782) [reply doc/user-guide/ggplot.md]: Just made the change

**Review by neelasha23 — DISMISSED** [review]

**Comment by bbeat2782** [comment]
@yafimvo 

I think the mention notification got missed, so I'm commenting again with the re-review request.

After clarifying the following with edublancas 
> mimic what ggplot in R does (this means that in some cases there will be whitespace if data is missing in certain data ranges).

do you think we still need to tackle the 3 plots you mentioned from your first comment?

**Comment by yafimvo** [comment]
> do you think we still need to tackle the 3 plots you mentioned from your first comment?

I'm not sure which 3 plots you are referring to.
Generally, as discussed, I agree that we should mimic what ggplot R does and treat it as our source of truth. We should ensure that our plots resemble the ones generated by ggplot R, making sure they are easily readable, and present accurate information.

**Review by yafimvo — APPROVED** [review]

**Review by yafimvo — APPROVED** [review]
I think the code looks good. Just make sure the cases we test match the ggplot R outputs.

**Review by edublancas — CHANGES_REQUESTED** [review]
I ran a few tests and seems like we're getting different results (left R, right JupysQL):

image

in R, the species with higher counts is Adelie, in ours is Chinstrap. Let's double check this.

it might be better to create some fake data to ensure we know beforehand how many observations there are.

**Comment by bbeat2782** [comment]
I will check it. 

@edublancas 
Just to make sure, when you said 
> in R, the species with higher counts is Adelie, in ours is Chinstrap

are you referring to their locations (which category is stacked on top of others)?

**Comment by edublancas** [comment]
> are you referring to their locations (which category is stacked on top of others)?

no, I mean the counts, the values in the Y axis

**Comment by bbeat2782** [comment]
@edublancas As you said, the binning in histogram works differently in R and our implementation. Thus, changing it will change lots of test histogram plots. Do you think it's better to change it here or in a different open PR?

**Comment by edublancas** [comment]
@AnirudhVIyer: this PR is having the issue that you're having.

@bbeat2782: the error that you're seeing is unrelated to your PR

I included a fix for this in https://github.com/ploomber/jupysql/pull/746, so once it's merged, you can rebase

**Comment by edublancas** [comment]
@bbeat2782 ok, let's keep this as is (once the tests pass, we can merge)

and let's tackle the shape of the histogram in a new PR (opened https://github.com/ploomber/jupysql/issues/751)

**Review by edublancas — APPROVED** [review]

**Pull request merged**